### PR TITLE
Added support for creating a prediction trajectory from a model cache

### PR DESCRIPTION
**Pull request by levmckinney**
This PR adds several new features that allow for easy integration with the [`TransformerLens`](https://github.com/neelnanda-io/TransformerLens) repo. 

Features:
* Support for prediction trajectories with a batch dimension.
* Support for creating prediction trajectories from model caches.
* Support for a new kind of log prob difference visualization that shows you how the delta between two tokens the `target` and the `antitarget` evolves across the trajectory.

## Recommending reading order
1.  If you are not familiar with the prediction trajectory class start with the `prediction_trajectories.ipynb` notebook. Note you may need to comment the thing out to get the plots to appear on your local system.
2. Read through the `combining_with_transformer_lens.ipynb` tutorial. This covers most of the new features added with some worked examples.
3. Look at the `prediction_trajectory.py`

## What I want feedback on
Does this interface with the transformer lens repo make sense? Is there anything obvious missing? Is the tutorial clear? Are there parts that need more explanation or might be misleading.

### CODE REVIEW

File: tuned_lens/plotting/trajectory_plotting.py [hunk 1]
```diff
@@ -1,24 +1,71 @@
...
     label_strings: NDArray[np.str_]
-    # (sequence_length) labels for the sequence dimension typically the input tokens.
-    sequence_labels: NDArray[np.str_]
-    # (n_layers x sequence_length x k) k entries to display when hovering over a cell.
-    # For example, the top k prediction from the lens at each layer.
+    # (n_layers x sequence_length x rows x cols) table of strings to display when
+    # hovering over a cell. For example, the top k prediction from the lens.
     hover_over_entries: Optional[NDArray[np.str_]] = None
 
+    def stride(self, stride: int) -> "TrajectoryLabels":
```
Comment (rhaps0dy) [inline tuned_lens/plotting/trajectory_plotting.py:25]: This function and `template_and_customdata` don't seem to be exercised by the tests. Are they in one of the notebooks above?

File: tuned_lens/plotting/prediction_trajectory.py [hunk 2]
```diff
@@ -1,8 +1,14 @@
...
 from dataclasses import dataclass
-from typing import Optional, Sequence, Union
+from typing import Literal, Optional, Sequence, Union
+
+try:
+    import transformer_lens as tl
+
+    _transformer_lens_available = True
+except ImportError:
+    _transformer_lens_available = False
```
Comment (rhaps0dy) [inline tuned_lens/plotting/prediction_trajectory.py:11]: I think when this line activates, line 94 of `model_surgery.py` will fail.

**Review by rhaps0dy — CHANGES_REQUESTED** [review]
- please fix the undefined variable error when `transformer_lens` is not present
- Could we test in CI what happens when `transformer_lens` is not present? Or make it a hard requirement.

- Could we test in CI that the tutorial notebooks don't crash? Just `import tutorials.prediction_trajectories` or something.

File: tuned_lens/model_surgery.py [hunk 3]
```diff
@@ -54,15 +62,38 @@ def assign_key_path(model: T, key_path: str, value: Any) -> Generator[T, None, N
...
 
 
-def get_final_norm(model: tr.PreTrainedModel) -> Norm:
+def get_final_norm(model: Model) -> Norm:
     """Get the final norm from a model.
 
     This isn't standardized across models, so this will need to be updated as
     we add new models.
     """
+    if _transformer_lens_available and isinstance(model, tl.HookedTransformer):
```
[the remark this answers is not in the document]
Comment (levmckinney) [reply tuned_lens/model_surgery.py:94]: Just tested this and It does not crash even when transformer lens is not installed.

File: docs/source/tutorials/combining_with_transformer_lens.ipynb [hunk 4]
```diff
@@ -0,0 +1,10508 @@
...
+   "source": [
+    "pred_traj_clean_logit.log_prob_diff(delta=True).clip(*log_prob_range).figure(title=\"Same as above but with the logit lens\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The above results mostly agree with the results in the paper. We see that the majority\n",
```
[the remark this answers is not in the document]
Comment (levmckinney) [reply docs/source/tutorials/combining_with_transformer_lens.ipynb:246]: Comment out the code in cell two. It's formatted for sphinx by default.

File: tuned_lens/model_surgery.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rhaps0dy) [reply tuned_lens/model_surgery.py:94]: Wow yeah, i guess the second clause doesn't get evaluated even to the level of checking if the variables are in scope. TIL!

**Review by rhaps0dy — APPROVED** [review]
Thanks for addressing all, very good job on the PR!!

File: tuned_lens/plotting/trajectory_plotting.py (hunk 1 above)
Comment (levmckinney) [reply tuned_lens/plotting/trajectory_plotting.py:25]: They are but I also added a unit test

**Pull request merged**